Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,361 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if !NET40
+// This HttpHandlerDiagnosticListener class is applicable only for .NET 4.5/4.6, and not for .NET core.
+// If you are making these changes, please test your changes manually via custom test applications.
+
+// NOTE!!!
+// This file is here temporarily. Once this functionality is in DiagnosticSource nuget, we should remove this class
+namespace Microsoft.ApplicationInsights.DependencyCollector.Implementation
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Net;
+    using System.Reflection;
+    using System.Reflection.Emit;
+
+    /// <summary>
+    /// A HttpHandlerDiagnosticListener is a DiagnosticListener for .NET 4.5/4.6 where HttpClient
+    /// doesn't have a DiagnosticListener built-in. This class is not used for .NET Core because
+    /// HttpClient in .NET Core already emits DiagnosticSource events. This class compensates for
+    /// that in .NET 4.5/4.6. HttpHandlerDiagnosticListener has no public constructor, but it has a
+    /// singleton style Initialize method to control its creation.
+    /// To use this, the application just needs to call HttpHandlerDiagnosticListener.Initialize(), and this
+    /// will register itself with the DiagnosticListener's all listeners collection.
+    /// </summary>
+    internal class HttpHandlerDiagnosticListener : DiagnosticListener
+    {
+        internal static HttpHandlerDiagnosticListener SingletonInstance = new HttpHandlerDiagnosticListener();
+
+#region private fields
+        private const string DiagnosticListenerName = "System.Net.Http.Desktop.V4";
+        private const string RequestWriteName = "System.Net.Http.Request";
+        private const string ResponseWriteName = "System.Net.Http.Response";
+        private const string ExceptionEventName = "System.Net.Http.Exception";
+
+        // Fields for reflection
+        private static FieldInfo connectionGroupListField;
+        private static Type connectionGroupType;
+        private static FieldInfo connectionListField;
+        private static Type connectionType;
+        private static FieldInfo writeListField;
+        private static Func<HttpWebRequest, HttpWebResponse> httpResponseAccessor;
+
+        // Fields for controlling initialization of the HttpHandlerDiagnosticListener singleton
+        private object initializationLock = new object();
+        private bool initialized = false;
+#endregion
+
+        /// <summary>
+        /// Prevents a default instance of the <see cref="HttpHandlerDiagnosticListener" /> class from being created.
+        /// Make sure every caller has to create this object via the Initialize method.
```

Comment:
Comment out of date? Seems that (internal) callers just access the singleton, right?